packages/stylist-macros/src/tokentree/parsed.rs

```diff
@@ -0,0 +1,535 @@
+        Ok(Self {
+            qualifiers,
+            qualifier_errors,
+        })
+    }
+}
+
+impl Parse for CssQualifiedRule {
+    fn parse(input: &ParseBuffer) -> ParseResult<Self> {
+        let qualifier = input.parse()?;
+        let scope = input.parse()?;
+        Ok(Self { qualifier, scope })
+    }
+}
+
+impl Parse for CssAtRule {
+    fn parse(input: &ParseBuffer) -> ParseResult<Self> {
+        let at = input.parse()?;
+        let name = input.parse::<CssIdent>()?;
+
+        // Consume all tokens till the next ';' or the next block
+        let mut component_iter = ComponentValueStream::from(input).peekable();
+        let mut prelude = vec![];
+        let mut errors = vec![];
+        if !["media", "supports"].contains(&name.stringify().as_str()) {
+            errors.push(ParseError::new_spanned(
+                &name,
+                format!("@-rule '{}' not supported", name.stringify()),
+            ));
+        }
+
+        let contents = loop {
+            if input.peek(token::Semi) {
+                let semi = input.parse()?;
+                break CssAtRuleContent::Empty(semi);
+            }
+            if input.peek(token::Brace) {
+                let scope = input.parse()?;
+                break CssAtRuleContent::Scope(scope);
+            }
+            let next_token = component_iter
+                .next()
+                .ok_or_else(|| input.error("AtRule: unexpected end of input"))??;
+            // unwrap okay, since we already peeked
+            prelude.push(next_token);
+        };
+
+        Ok(Self {
+            at,
+            name,
+            prelude,
+            contents,
+            errors,
+        })
+    }
+}
+
+impl ComponentValue {
+    fn maybe_to_attribute_name(self) -> Option<CssAttributeName> {
+        match self {
+            ComponentValue::Token(PreservedToken::Ident(i)) => {
+                Some(CssAttributeName::Identifier(i))
+            }
+            ComponentValue::Expr(expr) => Some(CssAttributeName::InjectedExpr(expr)),
+            _ => None,
+        }
+    }
+}
+
+impl ToTokens for CssScopeQualifier {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        for q in self.qualifiers.iter() {
+            q.to_tokens(tokens);
+        }
+    }
+}
+
+impl Default for CssScopeQualifier {
+    fn default() -> Self {
+        Self {
+            qualifiers: vec![],
+            qualifier_errors: vec![],
+        }
+    }
+}
+// =====================================================================
+// =====================================================================
+// Convert into output
+// =====================================================================
+// =====================================================================
+
+/// We want to normalize the input a bit. For that, we want to pretend that e.g.
+/// the sample input
+///
+/// ```css
+/// outer-attribute: some;
+/// foo-bar: zet;
+/// .nested {
+///     @media print {
```

Comment:
Just trying to mirror other libraries. It's sometimes very useful to have this kind of nesting in source code. I do not know how to best port this to the other parser.

Reply:
I guess maybe we can get away with this feature in this round. I don't think we can justify only one syntax flavour supports a feature and the other one doesn't.

If we remove it from this round of implementation of then we can also remove `fold_*` (which greatly reduces code complexity) as I think manipulating ast here is probably not the proper way to implement this feature if both sides introduces this syntax.

It's easy enough to write something like (for the moment):
```css
@media screen and (max-width: 500px) {
    .selector { color: red; }
}

@media screen and (max-width: 600px) {
    .selector { color: blue; }
}

@media screen and (max-width: 700px) {
    .selector { color: yellow; }
}
```

Reply:
Okay with me. I had planned to use it, e.g. to write

```css
&:hover {
    border: 1px solid ${color_secondary.main};
    background-color: ${to_hover(color_secondary.main)};
    @media (hover: none) {
        background-color: transparent;
    }
}
```
instead of
```css
&:hover {
    border: 1px solid ${color_secondary.main};
    background-color: ${to_hover(color_secondary.main)};
}
@media (hover: none) {
    &:hover { background-color: transparent; }
}
```

Reply:
Disallowed for now, easy enough to switch back on